### PR TITLE
🐛 Add +required tag and make members optional by default.

### PR DESCRIPTION
**Pull request by twiest**
- [x] Change the default behavior for members to be optional.
- [x] Allow members to be set with a +required comment flag.
- [x] Filter out +require comment flags from CRD description field.
- [x] Change testData to ensure +required is working as expected.

This addresses issue #137.

### CODE REVIEW

**Comment by damemi** [comment]
@sttts I think this would bring our fork in line with this repo

**Comment by damemi** [comment]
bump, any thoughts on this? I'm in favor of this behavior but it's basically the opposite of how crd generation works now

File: pkg/internal/codegen/parse/crd.go [hunk 1]
```diff
@@ -628,7 +628,7 @@ func (b *APIs) getMembers(t *types.Type, found sets.String) (map[string]v1beta1.
 			m, r := b.typeToJSONSchemaProps(member.Type, found, member.CommentLines, false)
 			members[name] = m
 			result[name] = r
-			if !strings.HasSuffix(strat, "omitempty") {
+			if hasRequired(member) {
```
Comment (droot) [inline pkg/internal/codegen/parse/crd.go]: Changing the default behavior in this case will break all the existing projects which is unacceptable.

**Comment by sttts** [comment]
We could make this package specific: look for doc.go upwards from the package of the current type until `// +optional=true` is found. Then `// +required` switch a field to being required.

**Comment by twiest** [comment]
@droot I'm looking at rebasing this PR for master, but I have a couple of questions first.

1) If `+required` is too much of a behavior change, would a PR that simply respected `+optional` as well as `omitempty` as discussed in issue #137 be acceptable?

2) In `master`, I don't see any unit tests for CRD generation. Am I missing them or do they not exist yet?

**Comment by sttts** [comment]
@twiest please sync with @damemi. We plan to implement a variant where you can switch the default of `+required/+optional` by putting the tag into 
  * doc.go
  * possibly the file (does that work in gengo?)
  * the type
  * the field

The default might stay as it is. But you can put `+optional` on the package and then `+required` is allowed to opt-in.

**Comment by damemi** [comment]
Also, according to @DirectXMan12 version 0.2 is a breaking release anyway, so if we *do* decide to switch the behavior this would be an appropriate time to do it

**Comment by twiest** [comment]
@sttts @damemi sounds great, thanks!

**Comment by DirectXMan12** [comment]
closing as per #211

**Comment by DirectXMan12** [comment]
thanks for your efforts though :-)

**Pull request closed without merging**